### PR TITLE
Collection data access APIs --> core

### PR DESCRIPTION
**Pull request by smurthas**
The collections can now (optionally) expose an api.js file in their src dir and it will be required in by core and add it's endpoints (GET only right now). This is intended to enable collection processes to turn themselves off more often (only need to run for data processing) and (hopefully) enabled better process priority control (nice down processing, while leaving core at normal priority).

This also add Common/node/collectionDataStore, which abstracts several common data store patterns in collections. Next step is to make some of the API methods common, but I'd rather merge (and deploy) this first.

This passes all tests, but they're aren't really a ton of collection tests. I'd be happy to write more, but if we are moving away from vows it feels like a waste (tell me I'm wrong!). I did auth all connectors to ensure data processing works properly and hit all API endpoints to make sure they work.

### CODE REVIEW

**Comment by smurthas** [comment]
I need at least one person to test this:
1. on **master**, clean me dir, connect all services
2. once all data is in, hit /state on all collections, record the result (mostly `count`)
3. repeat process from clean me dir on this branch
   - again, check /state, compare results -- should be the same as master!
   - check the logs closely for for any errors.
   - hit all collection API endpoints, listed below.

###Contacts
[/Me/contacts](http://localhost:8042/Me/contacts)
[/Me/contacts/](http://localhost:8042/Me/contacts/)
[/Me/contacts?limit=10&offset=25](http://localhost:8042/Me/contacts?limit=10&offset=25)
[/Me/contacts/state](http://localhost:8042/Me/contacts/state)
[/Me/contacts/since?id=<some-id>](http://localhost:8042/Me/contacts/since?id=<some-id)
[/Me/contacts/id/:someid](http://localhost:8042/Me/contacts/id/:someid)
[/Me/contacts/update](http://localhost:8042/Me/contacts/update) (and check logs again for errors!)

###Photos
[/Me/photos](http://localhost:8042/Me/photos)
[/Me/photos/](http://localhost:8042/Me/photos/)
[/Me/photos?limit=10&offset=25](http://localhost:8042/Me/photos?limit=10&offset=25)
[/Me/photos/state](http://localhost:8042/Me/photos/state)
[/Me/photos/since?id=<some-id>](http://localhost:8042/Me/photos/since?id=<some-id)
[/Me/photos/id/:someid](http://localhost:8042/Me/photos/id/:someid)
[/Me/photo/update](http://localhost:8042/Me/photo/update) (and check logs again for errors!)
[/Me/photos/image/:someid](http://localhost:8042/Me/photos/image/:someid)
[/Me/photos/thumbnail/:someid](http://localhost:8042/Me/photos/thumbnail/:someid)

###Places
[/Me/places](http://localhost:8042/Me/places)
[/Me/places/](http://localhost:8042/Me/places/)
[/Me/places?limit=10&offset=25](http://localhost:8042/Me/places?limit=10&offset=25)
[/Me/places/state](http://localhost:8042/Me/places/state)
[/Me/places/since?id=<some-id>](http://localhost:8042/Me/places/since?id=<some-id)
[/Me/places/id/:someid](http://localhost:8042/Me/places/id/:someid)
[/Me/places/update](http://localhost:8042/Me/places/update) (and check logs again for errors!)
[/Me/places/from/:network/:from](http://localhost:8042/Me/places/from/:network/:from) (eg /from/foursquare/someones4sqIdNum)
[/Me/places/from/geo/:network](http://localhost:8042/Me/places/from/geo/:network) (eg /geo/foursquare) (and check logs again for errors!)

###Links
[/Me/links](http://localhost:8042/Me/links)
[/Me/links/](http://localhost:8042/Me/links/)
[/Me/links?limit=10&offset=25](http://localhost:8042/Me/links?limit=10&offset=25)
[/Me/links/state](http://localhost:8042/Me/links/state)
[/Me/links/since?id=<some-id>](http://localhost:8042/Me/links/since?id=<some-id)
[/Me/links/id/:someid](http://localhost:8042/Me/links/id/:someid)
[/Me/links/encounters/:id](http://localhost:8042/Me/links/encounters/:id)
[/Me/links/update](http://localhost:8042/Me/links/update) (and check logs again for errors!)
[/Me/links/search?q=singly](http://localhost:8042/Me/links/search?q=singly)
[/Me/links/embed?url=http://www.youtube.com/watch?v=1bJxYRtGUM4](http://localhost:8042/Me/links/embed?url=http://www.youtube.com/watch?v=1bJxYRtGUM4)

File: Common/node/lservicemanager.js [hunk 1]
```diff
@@ -472,3 +477,19 @@ function checkForShutdown() {
     shuttingDown();
     shuttingDown = null;
 }
+
+exports.getCollectionApis = function() {
+  var collectionApis = {};
+  for (var i in serviceMap) {
+    if (serviceMap[i].type === 'collection') {
+      var modulePath = path.join(__dirname, '..', '..', serviceMap[i].srcdir, 'api.js');
```
Comment (quartzjer) [inline Common/node/lservicemanager.js]: Everywhere else I think we use lconfig.lockerDir here, would that be better than the ..'s?

Comment (smurthas) [inline Common/node/lservicemanager.js]: yep, thanks!

File: Ops/webservice.js [hunk 2]
```diff
@@ -202,6 +203,17 @@ locker.get('/core/:svcId/at', function(req, res) {
     res.end("true");
 });
 
+var collectionApis = serviceManager.getCollectionApis();
```
Comment (quartzjer) [inline Ops/webservice.js:205]: It seems like this should happen dynamically whenever a collection is upserted, it seems dangerous to depend on it existing at load-time? Also hard for local dev, and/or when collections are installed dynamically someday?

Comment (smurthas) [inline Ops/webservice.js:205]: yep, should definitely be on upsert! I'll change that.

**Comment by quartzjer** [comment]
Dude, kudos, this was quite the undertaking, moving a lot of logic around, thanks! It feels like a solid step towards some generalized collection functionality too, a pattern that could start to be applied to the data processing as well (I've had some ideas here, making this look more like synclets).

One caveat is this will increase the ram in the core process, and any bugs will bring the whole ship down, but the work @temas is doing with synclets using vm contexts as isolation might work perfectly here too.

Now to do some testing :)

File: Ops/webservice.js [hunk 3]
```diff
@@ -202,6 +203,17 @@ locker.get('/core/:svcId/at', function(req, res) {
...
+for(var i in collectionApis) {
+  locker._oldGet = locker.get;
+  locker.get = function(path, callback) {
+    return locker._oldGet('/Me/' + i + path, callback);
```
Comment (quartzjer) [inline Ops/webservice.js:209]: is there a way to plug in a try/catch here to make this safer?

Comment (smurthas) [inline Ops/webservice.js:209]: Good call. I'll add that as well!

On Feb 28, 2012, at 15:23, Jeremie Miller
reply@reply.github.com
wrote:

**Comment by quartzjer** [comment]
I ran this with every app, and I also tested the memory impact on my mac between the two for lockerd node process. By including the apis in the main process it only bumped up RSS by 2MB, not bad.

**Pull request merged**